### PR TITLE
Fixed app crash in rank widget issue (#2742) on config change.

### LINKED ISSUE #2742: Crash when device rotates in Rank widget
#### Software and hardware versions 
Collect v1.18.1, Android v7.0, device used : Moto m

#### Steps to reproduce the problem
1. Open ODK collect, download `All widgets` form.
2. Click `Fill Blank Form` and choose `All widgets`.
3. Go to Rank widget.
4. Click `Rank items` button.
5. Rotate device. (Make sure auto rotation setting is ON).
6. App crashes.

#### Expected behavior
No crash.

#### Other information 
Logs from android studio:
```
2018-11-26 19:26:23.064 17757-17757/? E/AndroidRuntime: FATAL EXCEPTION: main
    Process: org.odk.collect.android, PID: 17757
    java.lang.RuntimeException: Invalid query for Question prompt. Non-Question object at the form index
        at org.javarosa.form.api.FormEntryModel.getQuestionPrompt(FormEntryModel.java:170)
        at org.odk.collect.android.logic.FormController.getQuestionPrompt(FormController.java:864)
        at org.odk.collect.android.adapters.RankingListAdapter.onBindViewHolder(RankingListAdapter.java:61)
        at org.odk.collect.android.adapters.RankingListAdapter.onBindViewHolder(RankingListAdapter.java:41)
        at android.support.v7.widget.RecyclerView$Adapter.onBindViewHolder(RecyclerView.java:6781)
        at android.support.v7.widget.RecyclerView$Adapter.bindViewHolder(RecyclerView.java:6823)
        at android.support.v7.widget.RecyclerView$Recycler.tryBindViewHolderByDeadline(RecyclerView.java:5752)
        at android.support.v7.widget.RecyclerView$Recycler.tryGetViewHolderForPositionByDeadline(RecyclerView.java:6019)
        at android.support.v7.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:5858)
        at android.support.v7.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:5854)
        at android.support.v7.widget.LinearLayoutManager$LayoutState.next(LinearLayoutManager.java:2230)
        at android.support.v7.widget.LinearLayoutManager.layoutChunk(LinearLayoutManager.java:1557)
        at android.support.v7.widget.LinearLayoutManager.fill(LinearLayoutManager.java:1517)
        at android.support.v7.widget.LinearLayoutManager.onLayoutChildren(LinearLayoutManager.java:612)
        at android.support.v7.widget.RecyclerView.dispatchLayoutStep2(RecyclerView.java:3924)
        at android.support.v7.widget.RecyclerView.onMeasure(RecyclerView.java:3336)
        at android.view.View.measure(View.java:20110)
        at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6286)
        at android.widget.LinearLayout.measureChildBeforeLayout(LinearLayout.java:1705)
        at android.widget.LinearLayout.measureHorizontal(LinearLayout.java:1279)
        at android.widget.LinearLayout.onMeasure(LinearLayout.java:659)
        at android.view.View.measure(View.java:20110)
        at android.widget.ScrollView.measureChildWithMargins(ScrollView.java:1320)
        at android.widget.FrameLayout.onMeasure(FrameLayout.java:214)
        at android.widget.ScrollView.onMeasure(ScrollView.java:345)
        at android.view.View.measure(View.java:20110)
        at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6286)
        at android.widget.FrameLayout.onMeasure(FrameLayout.java:214)
        at android.view.View.measure(View.java:20110)
        at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6286)
        at android.widget.FrameLayout.onMeasure(FrameLayout.java:214)
        at android.view.View.measure(View.java:20110)
        at android.support.v7.widget.AlertDialogLayout.tryOnMeasure(AlertDialogLayout.java:134)
        at android.support.v7.widget.AlertDialogLayout.onMeasure(AlertDialogLayout.java:64)
        at android.view.View.measure(View.java:20110)
        at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6286)
        at android.widget.FrameLayout.onMeasure(FrameLayout.java:214)
        at android.support.v7.widget.ContentFrameLayout.onMeasure(ContentFrameLayout.java:143)
        at android.view.View.measure(View.java:20110)
        at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6286)
        at android.widget.FrameLayout.onMeasure(FrameLayout.java:214)
        at android.view.View.measure(View.java:20110)
        at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6286)
        at android.widget.FrameLayout.onMeasure(FrameLayout.java:214)
        at android.view.View.measure(View.java:20110)
        at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6286)
        at android.widget.LinearLayout.measureChildBeforeLayout(LinearLayout.java:1705)
        at android.widget.LinearLayout.measureVertical(LinearLayout.java:797)
        at android.widget.LinearLayout.onMeasure(LinearLayout.java:657)
        at android.view.View.measure(View.java:20110)
        at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:6286)
        at android.widget.FrameLayout.onMeasure(FrameLayout.java:214)
2018-11-26 19:26:23.064 17757-17757/? E/AndroidRuntime:     at com.android.internal.policy.DecorView.onMeasure(DecorView.java:714)
        at android.view.View.measure(View.java:20110)
        at android.view.ViewRootImpl.performMeasure(ViewRootImpl.java:2649)
        at android.view.ViewRootImpl.measureHierarchy(ViewRootImpl.java:1557)
        at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:1855)
        at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1471)
        at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:7009)
        at android.view.Choreographer$CallbackRecord.run(Choreographer.java:907)
        at android.view.Choreographer.doCallbacks(Choreographer.java:709)
        at android.view.Choreographer.doFrame(Choreographer.java:644)
        at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:893)
        at android.os.Handler.handleCallback(Handler.java:836)
        at android.os.Handler.dispatchMessage(Handler.java:103)
        at android.os.Looper.loop(Looper.java:203)
        at android.app.ActivityThread.main(ActivityThread.java:6293)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1065)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:926)

```

### PR DESCRIPTION
**Pull request by ajwad-shaikh**
Closes #2742 

#### What has been done to verify that this works as intended?
Tested on local device Moto G5 Plus

#### Why is this the best possible solution? Were any other approaches considered?
Adopts best practices for Saving and Retrieving Instances

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
None

#### Do we need any specific form for testing your changes? If so, please attach one.
Nope

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.
No

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

### CODE REVIEW

File: collect_app/src/main/java/org/odk/collect/android/fragments/dialogs/RankingWidgetDialog.java [hunk 1]
```diff
@@ -102,6 +102,7 @@ public Dialog onCreateDialog(Bundle savedInstanceState) {
     @Override
     public void onSaveInstanceState(Bundle outState) {
         outState.putSerializable(VALUES, (Serializable) rankingListAdapter.getValues());
+        outState.putSerializable(FORM_INDEX, rankingListAdapter.getFormIndex());
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/fragments/dialogs/RankingWidgetDialog.java]: Why don't you just use `formIndex` here? Then you don't need the method you implemented: `getFormIndex()`?

**Review by grzesiek2010 — CHANGES_REQUESTED** [review]

File: collect_app/src/main/java/org/odk/collect/android/fragments/dialogs/RankingWidgetDialog.java (hunk 1 above)
Comment (ajwad-shaikh) [reply collect_app/src/main/java/org/odk/collect/android/fragments/dialogs/RankingWidgetDialog.java]: @grzesiek2010 Took care of that. Thanks for pointing out 😄

Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/fragments/dialogs/RankingWidgetDialog.java]: Thanks!

**Comment by mmarciniak90** [comment]
**Tested with success**

Verified on Android: 4.1, 4.4, 5.1, 6.0, 7.0, 8.1
Crash does not occur when device orientation is changed on Rank Item dialog.

@opendatakit-bot label "behavior verified"
@opendatakit-bot unlabel "needs testing"

**Review by grzesiek2010 — APPROVED** [review]

**Pull request merged**